### ISSUE TITLE
Poor performance when initial value is near, but not exactly, 0

### ISSUE THREAD

**Issue by joshuahhh** [stevengj/nlopt#346]
I get extremely poor solving performance when the following conditions hold:
* solving with `NLOPT_LN_COBYLA` or `NLOPT_LN_SBPLX`
* an initial value is close to, but not exactly, 0
* system has at least two dimensions

Perhaps there's a bug! Or maybe I am misconfiguring my solver? (Though I am using `nlopt_set_xtol_abs1`, which I believe should not have any special behavior when values are near 0.)

My test code is below; output I get is at the bottom, in comments. I have nlopt 2.6.2 installed.

Thanks so much for taking a look!

```c
#include <nlopt.h>
#include <stdio.h>

// Squared distance to (3, 4)
double cost2D(unsigned n, const double *x, double *gradient, void *func_data) {
  double dx = x[0] - 3;
  double dy = x[1] - 4;
  return dx * dx + dy * dy;
}
void solve2D(nlopt_algorithm alg, double x_init, double y_init) {
  nlopt_opt opt = nlopt_create(alg, 2);
  nlopt_set_min_objective(opt, cost2D, NULL);
  nlopt_set_xtol_abs1(opt, 1e-6);
  double params[2];
  double obj;
  params[0] = x_init;
  params[1] = y_init;
  nlopt_result result = nlopt_optimize(opt, params, &obj);
  if (result < 0) {
    printf("Error: %i\n", result);
  } else {
    printf("Found solution %0.4f, %0.4f, with objective value %0.4f\n", params[0], params[1], obj);
  }
  nlopt_destroy(opt);
}

// Squared distance to 3
double cost1D(unsigned n, const double *x, double *gradient, void *func_data) {
  double dx = x[0] - 3;
  return dx * dx;
}
void solve1D(nlopt_algorithm alg, double x_init) {
  nlopt_opt opt = nlopt_create(alg, 1);
  nlopt_set_min_objective(opt, cost1D, NULL);
  nlopt_set_xtol_abs1(opt, 1e-6);
  double params[1];
  double obj;
  params[0] = x_init;
  nlopt_result result = nlopt_optimize(opt, params, &obj);
  if (result < 0) {
    printf("Error: %i\n", result);
  } else {
    printf("Found solution %0.4f, with objective value %0.4f\n", params[0], obj);
  }
  nlopt_destroy(opt);
}

int main() {
  solve2D(NLOPT_LN_COBYLA,    0, 1);  // Found solution 3.0000, 4.0000, with objective value 0.0000
  solve2D(NLOPT_LN_SBPLX,     0, 1);  // Found solution 3.0000, 4.0000, with objective value 0.0000
  solve2D(NLOPT_LN_COBYLA, 1E-3, 1);  // Found solution 1.9658, 4.0010, with objective value 1.0695
  solve2D(NLOPT_LN_SBPLX,  1E-3, 1);  // Found solution 3.0000, 4.0000, with objective value 0.0000
  solve2D(NLOPT_LN_COBYLA, 1E-8, 1);  // Found solution 0.0000, 4.5694, with objective value 9.3243
  solve2D(NLOPT_LN_SBPLX,  1E-8, 1);  // Found solution 0.0000, 4.0000, with objective value 9.0000

  solve1D(NLOPT_LN_COBYLA,    0);  // Found solution 3.0000, with objective value 0.0000
  solve1D(NLOPT_LN_SBPLX,     0);  // Found solution 3.0000, with objective value 0.0000
  solve1D(NLOPT_LN_COBYLA, 1E-3);  // Found solution 3.0000, with objective value 0.0000
  solve1D(NLOPT_LN_SBPLX,  1E-3);  // Found solution 3.0000, with objective value 0.0000
  solve1D(NLOPT_LN_COBYLA, 1E-8);  // Found solution 3.0000, with objective value 0.0000
  solve1D(NLOPT_LN_SBPLX,  1E-8);  // Found solution 3.0000, with objective value 0.0000
}
```

**Comment by stevengj** [comment]
You may need to set the initial step size manually (https://nlopt.readthedocs.io/en/latest/NLopt_Reference/#initial-step-size), as otherwise it uses some heuristics that may be failing in your case.

**Comment by joshuahhh** [comment]
You are absolutely right. Adding `nlopt_set_initial_step1(opt, 1);` (or pretty much any other step size) fixed it right up. Wow – thanks!

Ideally the heuristics could be fixed up a bit for this case, but for now, I submitted a PR (#347) to add a little warning to the docs. I hope this is helpful; let me know if there's any issues with it. Thanks again for your speedy response!

**Comment by stevengj** [comment]
Closed by #347.
